### DISCUSSION TITLE
will axmol fasttilemap  support more than 1 tileset per layer

### DISCUSSION THREAD

**Discussion by jcy1001** [axmolengine/axmol#1308] [General]
will axmol fasttilemap  support more than 1 tileset per layer?
now it cant,
i use this test tmx , one layer use 2 tileset image ,but fasttilemap not support display correct
[test.zip](https://github.com/axmolengine/axmol/files/12444953/test.zip)
image
demo use cpp_tests

**Comment by jcy1001** [comment]
![image](https://github.com/axmolengine/axmol/assets/24226340/9608e2de-2183-4f5a-8d95-94336d71da8d)

**Comment by jcy1001** [comment]
i found fasttilemap commet 
TMXTilesetInfo* FastTMXTiledMap::tilesetForLayer(TMXLayerInfo* layerInfo, TMXMapInfo* mapInfo)
{
    Vec2 size      = layerInfo->_layerSize;
    auto& tilesets = mapInfo->getTilesets();

    for (auto iter = tilesets.crbegin(), iterCrend = tilesets.crend(); iter != iterCrend; ++iter)
    {
        TMXTilesetInfo* tilesetInfo = *iter;
        if (tilesetInfo)
        {
            for (int y = 0; y < size.height; y++)
            {
                for (int x = 0; x < size.width; x++)
                {
                    uint32_t pos = static_cast<uint32_t>(x + size.width * y);
                    uint32_t gid = layerInfo->_tiles[pos];

                    // gid are stored in little endian.
                    // if host is big endian, then swap
                    // if( o == CFByteOrderBigEndian )
                    //    gid = CFSwapInt32( gid );
                    /* We support little endian.*/

                    // FIXME: gid == 0 --> empty tile
                    if (gid != 0)
                    {
                        // Optimization: quick return
                        // if the layer is invalid (more than 1 tileset per layer) an CCAssert will be thrown later
                        if ((gid & kTMXFlippedMask) >= static_cast<uint32_t>(tilesetInfo->_firstGid))
                        {
                            return tilesetInfo;
                        }
                    }
                }
            }
        }
    }

    // If all the tiles are 0, return empty tileset
    AXLOG("cocos2d: Warning: TMX Layer '%s' has no tiles", layerInfo->_name.c_str());
    return nullptr;
}

**Reply by halx99** [reply]
I guess it's related to https://github.com/axmolengine/axmol/pull/1142
